MSTR-258 : s3, huggingface auto upload, Back Translation Data Augmentation, Early Stopping 로직 추가

ekzm8523 opened the pull request:
### Issue Number

close: MSTR-253, MSTR-254, MSTR-257, MSTR-258, MSTR-259

### 작업 내역

- [x] S3, Huggingface에 자동으로 Best Model이 Upload 가능하도록 하는 로직 추가 (Hydra config로 설정 가능)
- [x] Wandb를 통해 실험 추적 기능 추가
    - ![스크린샷 2022-09-19 오전 3 01 43](https://user-images.githubusercontent.com/67869514/190921789-07f65e2c-4ddc-4fe9-b539-02e51420fc38.png)
- [x] Train, Val, Test 구조에서 Test file은 고정하고 Split은 스크립트 형태로 변경
    - 실험 코드 상에서 split하는 부분은 사라졌습니다.
    - Competition에 익숙한 나머지 잘못된 설계를 한 부분을 고친 것 입니다.
- [x] Overfitting을 방지하도록 Early Stopping 기능을 추가하였습니다.
- [x] threshold가 없기 때문에 auc metric를 삭제하였고 Joint Goal Accuracy(JGA)를 추가 할 예정입니다.
    - JGA란 채점시 유저 답변에 4가지 채점기준이 있었다면 전부 라벨과 일치해야 맞았다고 처리하는 metric
- [x] Back Translation Data Augmentation 코드를 추가하였습니다.
- [x] 전체적으로 대규모 리팩토링이 있었습니다..

### 작업 유형


- [x] 신규 기능 추가
- [x] 버그 수정
- [x] 리펙토링
- [x] 문서 업데이트

### PR 특이 사항
- 이제 실험 일지를 Confluence와 세부 사항은 Private notion에 정리 할 예정이고 다음 링크에서 보실 수 있습니다.
    - [confluence](https://swm-monstera.atlassian.net/wiki/spaces/WPLCR/pages/51249188)
    - [notion](https://circular-harpymimus-2dc.notion.site/7f5b2ce64f9f43809d6e58427cb22000?v=0dacf7bd09af4a25b5c5a724f874524d)
- 또한 위 스크린샷에 있는 wandb도 팀 초대가 가능하여 실험 진행사항을 더 세밀하고 보고싶다면 초대를 해드리겠습니다.
- Back Translation Data Augmentation(BTDA)을 위해 Naver Developer papago API를 사용하였지만 하루 10000자의 번역만 가능하기 때문에 Celenium 크롤러를 이용해서 BTDA csv를 생성하였습니다.
- 실험과 코딩을 병행하다보니 PR의 몸집이 많이 무거워져서 죄송합니다.. 하지만 현조에 비한다면... 양반.. 화이팅

ekzm8523 committed:
:sparkles: Upload model to huggingface and s3 logic

upload model to huggingface and s3 logic

ekzm8523 committed:
:mute: remove config and log file

remove config add log file

ekzm8523 committed:
:sparkles: add wandb

add wandb

ekzm8523 committed:
:recycle: Add auroc metric and remove validation code

add auroc metric and remove validation code

ekzm8523 committed:
:technologist: integrate test and validation dataset

integrate test and validation dataset

ekzm8523 committed:
:sparkles: Add Early Stopping

add early stopping

ekzm8523 committed:
:recycle: Remove auc metric and refactoring code

remove auc metric and refactoring code

ekzm8523 committed:
:recycle: Remove validation split code and integrate test dataset

remove validation split code and integrate test dataset

ekzm8523 committed:
:technologist: add description input and change log shape

add description input and change log shape

ekzm8523 committed:
:sparkles: Add Back translation augmentation feature

add back translation augmentation feature

ekzm8523 committed:
:sparkles: Remove using papago developer api and Add celenium croller for papago

remove using papago developer api and add celenium croller for papago

kshired commented on core/config.py:
```diff
@@ -1,8 +1,40 @@
+    try:
+        get_secret_value_response = secret_manager.get_secret_value(SecretId=secret_name)
+    except ClientError as e:
+        log.error(e)
+        if e.response["Error"]["Code"] == "DecryptionFailureException":
+            raise e
+        elif e.response["Error"]["Code"] == "InternalServiceErrorException":
+            raise e
+        elif e.response["Error"]["Code"] == "InvalidParameterException":
+            raise e
+        elif e.response["Error"]["Code"] == "InvalidRequestException":
+            raise e
+        elif e.response["Error"]["Code"] == "ResourceNotFoundException":
+            raise e
```
error code를 확인하고, raise만 하는데 나중에 추가 할 예정인건가요?

ekzm8523 replied:
서비스 코드가 아니라서 로그만 띄워주고 Exception 정의는 따로 하지 않았습니다!

kshired commented on prompt_tuning/dataset.py:
```diff
@@ -90,53 +89,35 @@ def _get_preprocessed_dataset(self) -> List:
-
-        train_dataset, val_dataset, test_dataset = [], [], []
+        train_dataset = []
         for i, data in enumerate(every_dataset):
```
```suggestion
        for data in every_dataset:
```

kshired commented on prompt_tuning/dataset.py:
```diff
@@ -167,39 +148,14 @@ def _load_labeled_dataset(csv_path: str = "../static/labeled_dataset.csv") -> Li
-
-        train_dataset, val_dataset, test_dataset = [], [], []
-
+    def _convert_to_prompt_input_dataset(self) -> List[InputExample]:
```
얘는 위의 `class PromptLoader` 의 함수와 일치하는데, 같은 동작을해서 그런건가요?
이런애들은 java나 kotlin에서 interface로 상속하는 것 처럼 안되나..

ekzm8523 replied:
`class PromptNliDataModule`를 말씀하시는 것 같은데 내부동작이 살짝 다르지만 하고자 하는게 같아서 interface로 상속을 할 수 있기는 해요!! ABCMeta class라는걸 사용하거나 구현하지 않았을 때 raise되도록 해서 java의 interface를 구현할 수는 있지만 interface화 하기엔 하나의 메서드만 겹쳐서.. 하나의 클래스가 늘어나거나 조금 더 겹치게 된다면 고려해보겠습니다!

kshired commented on prompt_tuning/utils.py:
```diff
@@ -20,15 +26,102 @@ def seed_everything(seed):
-                f"accuracy: {accuracy_score / step} "
-                f"f1_score: {f1_score / step}"
+        if test_type == "train":
+            log_string = f"[{test_type}] Epoch {epoch} loss: {loss / step} "
+            log.info(log_string)
```
```suggestion
            log.info(f"[{test_type}] Epoch {epoch} loss: {loss / step} ")
```

굳이 다시 사용되는 변수가 아니라면, 리터럴로 써도 될 거 같아요.

kshired replied:
else 이하도 마찬가지라고 생각하는데, else이하는 string이 길어져서 뺀거면 OK!

ekzm8523 replied:
적용하겠습니다~!

kshired commented on prompt_tuning/utils.py:
```diff
@@ -20,15 +26,102 @@ def seed_everything(seed):
+
+    log.info(f"{bucket}/{folder}/{model_name}에 모델 업로드중")
+    s3.upload_file(local_path, bucket, f"{folder}/{model_name}")
+    log.info("모델 업로드 완료")
```
위에 log로 model name에 업로드 중입니다가 있긴 한데, 한 줄에 같이 어떤 모델인지 알려줘도 좋을 듯!

ekzm8523 replied:
어떤 모델인지는 config 설명에 다 적혀있고 여러 모델들을 사용하는게 아니라 따로 남겨놓진 않았어요!
만약 어떤 모델인지 알려주려면 파라미터를 추가해야하는데 로깅을 위해 파라미터를 추가하는건 뭔가 내키지 않네요 ㅠㅠ

kshired commented on static/papago.py:
```diff
@@ -0,0 +1,66 @@
+driver.maximize_window()
+
+
+def translate_with_papago(text: str, source_language: str, target_language: str) -> str:
+    """
+    trans_lang에 넣는 파라미터 값:
+    'en' -> 영어
+    'ja&hn=0' -> 일본어
+    'zh-CN' -> 중국어(간체)
+    """
+    try:
+        driver.get(f"https://papago.naver.com/?sk={source_language}&tk={target_language}&st={text}")
+        time.sleep(2.5)
+        translated_text = driver.find_element(By.XPATH, r'//*[@id="txtTarget"]').text
+        return translated_text
+    except:  # noqa
+        driver.get(f"https://papago.naver.com/?sk={source_language}&tk={target_language}")
+        driver.find_element(By.XPATH, r'//*[@id="txtSource"]').send_keys(text)
+        time.sleep(2.5)
+        translated_text = driver.find_element(By.XPATH, r'//*[@id="txtTarget"]').text
+        return translated_text
```
엥 번역 때, 파파고를 쓰는 것이였군요. 근데.. 이거 실 서비스 때 쓰는게 아니라 학습 때 쓰는 코드인거죠?

ekzm8523 replied:
서비스도 학습도 아니고 학습 전에 csv 추출하는데 사용됩니다!!

kshired commented on static/papago.py:
```diff
@@ -0,0 +1,66 @@
+
+def back_translation_augmentation(train_csv_path: str, start_idx: int = 0) -> None:
+    df = pd.read_csv(train_csv_path)
+    print(f"previous data size : {len(df)}")
```
얘는 왜 logger 말고 print일까요!

ekzm8523 replied:
학습에서 사용하는 것과 스크립트 형태로 일시적이게 사용하는 것을 분리하기 위해 단순히 print로 찍었습니다!!
로그에 남겨야 할 코드들이 아니라서 print를 찍었습니다~

kshired commented on static/papago.py:
```diff
@@ -0,0 +1,66 @@
+            back_translated_user_answer = back_translate(new_df.iloc[idx].user_answer)
+            new_df.iloc[idx].user_answer = back_translated_user_answer
+    except:  # noqa
+        print(f"에러가 발생했으니 {idx}부터 다시 시작하세요")
```
학습때 쓰는게 맞다면, 굳이라고 생각하지만.. 서비스에 사용되는 모델이라면 슬랙으로 보내봅시다!

ekzm8523 replied:
서비스도 학습도 아니에요.. ㅠㅠ

kshired commented on static/papago.py:
```diff
@@ -0,0 +1,66 @@
+from selenium import webdriver
+from selenium.webdriver.common.by import By
+
+driver = webdriver.Chrome()
+driver.maximize_window()
+
```
잉 파파고 API를 쓰는게 아니라 크롤링형태로 하나보네요..

API가 있다면, API를 쓰는게 최고일 것 같고.. 그게 아니라면 selenium을 안쓰고 requests 모듈 같은 것을 이용해서 정적으로 파싱할 수 있는 방안을 찾는게 좋아요.

크로미움 생각보다 꽤 무거워서.. 올렸다 내렸다하는데 이게 속도 꽤 잡아먹을거예요.

ekzm8523 replied:
맞아요.. 진짜 무겁더라구요.. 그래서 이 코드의 위치를 static폴더에 넣은 이유가 실험중에 사용될 수 없다고 판단해서 단순 스크립트 형태로 사용하기 위해 static 폴더에 놨어요!! 현재는 필요한 번역데이터를 다 뽑았기 때문에 더 대용량을 해야하고 속도가 필요하다면 고려해볼게요!!

kshired reviewed (commented):
comment 확인해주세요~

++ `.gitignore` 에 `.DS_Store` 추가해서, 이 파일은 포함되지 않도록해주세요!


### Tip : 전역으로 `.DS_Store` 무시하게하기

1. `.gitignore_global` 파일 생성 및 `.DS_Store` 추가
`echo .DS_Store >> ~/.gitignore_global`

2. 모든 repository에게 이걸 global하게 exclude하라고 config.
`git config --global core.excludesfile ~/.gitignore_global`

ekzm8523 committed:
Update prompt_tuning/dataset.py

Co-authored-by: SEONGILKIM <shiroed1211@gmail.com>

kshired reviewed (approved):
LGTM

ekzm8523 committed:
:recycle: change string to literal

change string to literal

ekzm8523 committed:
:sparkles: add git ignore golbal

add git ignore global

ekzm8523 committed:
Merge branch 'feat/258' of https://github.com/SW13-Monstera/AI-research into feat/258

ekzm8523 committed:
:sparkles: Add joint goal accuracy metric and evaluator class

add joint goal accuracy metric and evaluator class and test and train print split

ekzm8523 committed:
Update .gitignore_global

ekzm8523 committed:
:bug: fix bug

fix bug

ekzm8523 committed:
Merge branch 'feat/258' of https://github.com/SW13-Monstera/AI-research into feat/258

ekzm8523 committed:
:adhesive_bandage: split early stopping standard

split early stopping standard

ekzm8523 committed:
Merge branch 'dev' into feat/258
